Intro to CSS lesson: Reorganise "Intro to CSS" to encourage earlier hands-on practice (#31255)
TheOdinProject/curriculum: foundations/html_css/css_foundations/intro_to_css.md

This section contains a general overview of topics that you will learn in this lesson.

- Adding styles to HTML with CSS.
- How to use the `class` and `id` attributes.
- How different CSS selectors work.

### Adding CSS to HTML

Before learning how to write CSS rules, let's first link a CSS file to our `index.html` so you'll be able to experiment with every example throughout this lesson.

There are three ways to add CSS to HTML:

- External CSS
- Internal CSS
- Inline CSS

We'll focus on external CSS for now because it's the method you'll use most often. The two remaining methods will be covered later in this lesson.

#### External CSS

> External CSS is the most common method you will come across, and it involves creating a separate file for the CSS and linking it inside of an HTML's opening and closing `<head>` tags with a void `<link>` element:

Reviewer: Reduces repetition with the next paragraph

```suggestion
External CSS is the most common method you will come across, and it involves creating a separate file for CSS and linking it inside the HTML.
```

```html
<!-- index.html -->

<head>
  <link rel="stylesheet" href="styles.css">
</head>
```

> First, we add a void `<link>` element inside of the opening and closing `<head>` tags of the HTML file. The `href` attribute is the location of the CSS file, either an absolute URL or, what you'll be utilizing, a URL relative to the location of the HTML file. In our example above, we are assuming both files are located in the same directory. The `rel` attribute is required, and it specifies the relationship between the HTML file and the linked file.

Reviewer: Slight rephrasing as well as correcting "relative URL" since the path we use is not a URL.

```suggestion
First, we add a void `<link>` element inside of the opening and closing `<head>` tags of the HTML file. The `href` attribute is the location of the CSS file, in our case a path to it that's relative to the location of the HTML file (but could also be something like a URL to a stylesheet hosted somewhere else). In our example above, we are assuming both files are located in the same directory. The `rel` attribute is required, and it specifies the relationship between the HTML file and the linked file.
```

- Understand how to use the class and ID attributes.
- Add styles to specific elements using the correct selectors.

### Adding CSS to HTML

> Before learning how to write CSS rules, let's first link a CSS file to our `index.html` so you'll be able to experiment with every example throughout this lesson.

Reviewer: I don't think this is the right place to put this sentence (see a couple of comments below).

> There are three ways to add CSS to HTML:
> 
> - External CSS
> - Internal CSS
> - Inline CSS

Reviewer: Putting the "focus" bit in this section makes more sense.

```suggestion
There are three ways to add CSS to HTML:

- External CSS
- Internal CSS
- Inline CSS

We'll focus on external CSS for now because it's the method you'll use most often. The two remaining methods will be covered later in this lesson.
```

#### External CSS

External CSS is the most common method you will come across, and it involves creating a separate file for the CSS and linking it inside of an HTML's opening and closing `<head>` tags with a void `<link>` element:

```html
<!-- index.html -->

<head>
  <link rel="stylesheet" href="styles.css">
</head>
```

First, we add a void `<link>` element inside of the opening and closing `<head>` tags of the HTML file. The `href` attribute is the location of the CSS file, either an absolute URL or, what you'll be utilizing, a URL relative to the location of the HTML file. In our example above, we are assuming both files are located in the same directory. The `rel` attribute is required, and it specifies the relationship between the HTML file and the linked file.

Then, we create an empty `styles.css` file in the same directory as your `index.html` file. We'll start adding CSS rules to this file in the next section.

> A note on file names: `styles.css` is just what we went with as the file name here. You can name the file whatever you want as long as the file type is `.css`, though "style" or "styles" is most commonly used.

Reviewer: While we're at it, I think a [note box](https://github.com/TheOdinProject/curriculum/blob/main/LAYOUT_STYLE_GUIDE.md#note-boxes) would be great for this bit:
```suggestion
<div class="lesson-note" markdown="1">

#### CSS file names

`styles.css` is just what we went with as the file name here. You could name the file whatever you want, as long as the file type is `.css`; "style" or "styles" is just what's most commonly used.

</div>
```

A couple of the pros to this method are:

1. It keeps our HTML and CSS separated, which results in the HTML file being smaller and making things look cleaner.
1. We only need to edit the CSS in *one* place, which is especially handy for websites with many pages that all share similar styles.

> We'll focus on external CSS for now because it's the method you'll use most often. The two remaining methods will be covered later in this lesson.

Reviewer: This place makes more sense for the file creation instruction. Instructions for `index.html` are required as well.
```suggestion
To help us with learning how to write CSS rules, let's first create an `index.html` file (with the usual boilerplate) as well as a `styles.css` file in the same directory, then link both files together as described above.
```

Selectors refer to the HTML elements to which CSS rules apply; they're what is actually being "selected" for each rule. The following subsections don't cover every selector available, but they're by far the most common and the ones you should get comfortable using first.

> As you learn each selector, try writing it in your `styles.css` file, save the file, and refresh your browser to see the results.

Reviewer: Just a rephrasing suggestion. The original grammar was a little off but I think a full rephrase makes it flow a touch better.

```suggestion
As you go through each example, play around with them to see the results by adding both the HTML and CSS to their respective files.
```

These properties work in conjunction with the height and width attributes in the HTML. It’s best to include both of these properties and the HTML attributes for image elements, even if you don’t plan on adjusting the values from the image file’s original ones. When these values aren’t included, if an image takes longer to load than the rest of the page contents, it won’t take up any space on the page at first but will suddenly cause a drastic shift of the other page contents once it does load in. Explicitly stating a `height` and `width` prevents this from happening, as space will be “reserved” on the page and appear blank until the image loads.

> ### Additional Methods for Adding CSS to HTML

Reviewer: Our headings should be in [sentence case](https://github.com/TheOdinProject/curriculum/blob/main/LAYOUT_STYLE_GUIDE.md#case).

```suggestion
### Additional methods for adding CSS to HTML
```

> Now that we've learned some basic syntax, there are two additional methods to add CSS to HTML, along with the external method we learned earlier in this lesson.

Reviewer: Definitely a nit but "along with", at a glance, could sound like we'll now talk about the two additional methods *plus* more on the earlier external method.
```suggestion
Now that we've learned some basic syntax, there are two additional methods to add CSS to HTML (on top of the external method we learned earlier in this lesson).
```
